proposals/lip-0055.md

```diff
@@ -660,7 +660,12 @@ The signature is verified by calling the `verifyBlockSignature` function. This f
 
 ```python
 def verifyBlockSignature():
-  generatorKey = validators.getValidatorKeys(block.header.generatorAddress).generatorKey
+  validators = getBFTParameters(block.header.height).validators
+  keys = [v.generatorKey for v in validators if v.address == block.header.generatorAddress]
+  if len(keys) == 0:
```

Comment:
Duplications shouldn't happen, but we could check like

```python
if len(keys) != 1:
```

to be completely covered.

Reply:
Yes, we could. But I wanted to follow the [implementation](https://github.com/LiskHQ/lisk-sdk/blob/f45f5fa24858b133841953cbe61286531391e7ff/framework/src/engine/consensus/consensus.ts#L737-L747) also with regard to the exception message. If we change it, then it may also be required to change the implementation if we want to be strict.

Reply:
I would also ensure here that the length is exactly 1, even if that means the implementation has to change.

Reply:
I'm not very happy with it as we are adding some (probably negligible) unnecessary computation steps to detect a bug and not to detect some valid states, but I'll do it. Tagging also @shuse2.

Reply:
if duplication doesn't happen, i think no point to add this?
The implementation assumes there is no duplication, and if it happens it should be checked when it's inserted?

Reply:
In my understanding `getBFTParameters` is done in the engine (LIP58), so that it is not customizable. If we can guarantee that no duplication can happen, I agree it's fine to just check `len(keys) == 0`.

Reply:
We decided to keep this PR as it is.

@shuse2 we nevertheless found that there is currently not check that ensures that there is no duplication when data is set in consensus. @janhack will create a separate PR in which the checks are added.

Reply:
Here is the issue that I created https://github.com/LiskHQ/lips/issues/279.